src/core.ts

```diff
@@ -1,62 +1,137 @@
+  Storage?: StorageConfig;
+  API?: APIConfig;
+};
 
 export type StorageConfig = {
-    bucket: string,
-    region: string, // or scope it down to actual regions
+  bucket: string;
+  region: string; // or scope it down to actual regions
 };
 
 export type APIConfig = {};
-export type AuthConfig = StrictUnion<UserPoolConfig | IdentityPoolConfig | UserPoolAndIdentityPoolConfig>;
+export type AuthConfig = StrictUnion<
+  UserPoolConfig | IdentityPoolConfig | UserPoolAndIdentityPoolConfig
+>;
 
 type UserPoolConfig = {
-    userPoolId: string,
-    userPoolClientId: string,
-    signUpVerificationMethod?: 'code' | 'link',
-    oauth?: {
-        domain: string,
-        scope: UserPoolScopes,
-        redirectSignIn: string,
-        redirectSignOut: string,
-        responseType: 'code' | 'token'
-    }
-}
+  userPoolId: string;
+  userPoolClientId: string;
+  signUpVerificationMethod?: "code" | "link";
+  oauth?: {
+    domain: string;
+    scope: UserPoolScopes;
+    redirectSignIn: string;
+    redirectSignOut: string;
+    responseType: "code" | "token";
+  };
+};
 
-type UserPoolScope = 'phone' | 'email' | 'profile' | 'openid' | 'aws.cognito.signin.user.admin';
+type UserPoolScope =
+  | "phone"
+  | "email"
+  | "profile"
+  | "openid"
+  | "aws.cognito.signin.user.admin";
 
 type UserPoolScopes = UserPoolScope[];
 
 type IdentityPoolConfig = {
-    identityPoolId: string,
+  identityPoolId: string;
+};
+
+type UserPoolAndIdentityPoolConfig = UserPoolConfig & IdentityPoolConfig;
+
+declare class Amplify {
+  configure<AmplifyConf extends AmplifyConfigure>(
+    config: AmplifyConfigure
+  ): void;
 }
 
-type UserPoolAndIdentityPoolConfig =  UserPoolConfig & IdentityPoolConfig
+type AmplifyChannel = "auth" | "api" | "storage" | "analytics";
 
-type AmplifyType = {
-    configure: (config: AmplifyConfigure) => void,
+type HubCapsule<
+  Channel extends AmplifyChannel,
+  Payload extends HubPayload = HubPayload
+> = {
+  channel: Channel;
+  payload: Payload;
+  source: string;
+  patternInfo?: string[];
+};
 
-}
+type SignInResult = {
+  isSignedIn: boolean;
+};
+
+type SignUpResult = {
+  isSignedUp: boolean;
+};
 
-export const Amplify: AmplifyType = {
-    configure: function (config: AmplifyConfigure): void {
-        throw new Error("Function not implemented.")
-    }
+type AuthError = {};
 
-}
+type AuthEvenDataMap = {
+  signIn: SignInResult;
+  signUp: SignUpResult;
+  signIn_failure: AuthError;
+  signUp_failure: AuthError;
+};
 
-export function httpClient(input: HTTPClientInput): Promise<Response> {
-    throw new Error("Function not impemented");
+type HubCallback<
+  Channel extends AmplifyChannel,
+  Event extends string = string,
+  Data extends any = any,
+  Payload extends HubPayload = HubPayload<Event, Data>
+> = (capsule: HubCapsule<Channel, Payload>) => void;
+type HubPayload<Event extends string = string, Data extends any = any> = {
+  event: Event;
+  data?: Data;
+  message?: string;
+};
+
+type HubCallbackMap<
```

Comment:
I am not sure the callback do the Event Mapping properly, I had to do a cast for the `data` depending on which event I was comparing. I think this might be solved if the HubPayload is an Strict union of all the different possible events, in case there is a new is matter of add that to the list.

What do you think?

Reply:
To my knowledge , I don't think a strict union would solve that. The issue is that the DS that contains `event` and `data` is at the same level, so `event` key can take multiple events and `data` can take multiple data objects. TS is not able no infer the right `data` for the right event in the switch statement. Another solution would be to nest the `data` object inside a given `event` .

Reply:
I think what we're missing here is an explicit relationship from event ID to the associated payload that can be used in the switch statement to infer to appropriate payload type. I found this example which seems like it could be useful: https://stackblitz.com/edit/switch-with-types?file=types.d.ts

Reply:
Thanks for the feedback. Now it is able to map the right data